src/xmpp/message.c

```diff
@@ -1307,17 +1234,34 @@ _handle_chat(xmpp_stanza_t *const stanza, gboolean is_mam)
     }
     Jid *jid = jid_create(from);
 
+    Jid *to_jid = NULL;
+    const gchar *to = xmpp_stanza_get_to(stanza);
+    if (to) {
+        to_jid = jid_create(to);
+    }
+
     // private message from chat room use full jid (room/nick)
     if (muc_active(jid->barejid)) {
         _handle_muc_private_message(stanza);
         jid_destroy(jid);
+        jid_destroy(to_jid);
         return;
     }
 
     // standard chat message, use jid without resource
     ProfMessage *message = message_init();
-    message->from_jid = jid;
     message->is_mam = is_mam;
+    message->from_jid = jid;
+    if (to_jid) {
+        message->to_jid = to_jid;
+    } else {
+        if (is_carbon) {
+            // happens when receive a carbon of a self sent message
+            // really? maybe some servers do this, but it's not required.
```

Comment:
I know this comment existed before... It's talking to itself?

Maybe rewrite it and elaborate?

Reply:
First line was present earlier, later then I added the second line because I have doubts about it. I still need to check it in more detail. So far it contains all I know :)